### ISSUE TITLE
Fast(er) UniFrac

### ISSUE THREAD

**Issue by defleury** [joey711/phyloseq#524]
Dear Joey!

I recently played around a bit with your code for ```fastUniFrac``` calculations. I was running into a wall when trying to compute pairwise distances for a ~4,000 samples * ~30,000 OTUs table. It seemed that UniFrac from phyloseq was quite slow and eating uneven amounts of memory per forked parallel process. So I gave it a shot and tried to tweak your code a little to run on my bigger problem.

I found a couple of things to speed up (in particular in the pre-processing for weighted UniFrac). Here's the code which I used eventually:

```R
#Load packages for parallel processing
require("foreach");
require("bigmemory");
library("doMC", quietly=T);
#Register cluster
registerDoMC(cores=use.cores);
########################

########################
#Get data
sp.o.t <- ot;                              #where ot is a previously built sparse Matrix object
tree <- phy_tree(my.ps);
#Get N x 2 matrix of pairwise combinations of samples
spn <- combn(samples, 2, simplify=FALSE);
########################

########################
#Build pre-requisite matrices
########################
# Create a list of descendants, starting from the first internal node (root)
descList <- prop.part(tree, check.labels = FALSE)
# Add the terminal edge descendants (tips). By definition, can only have one descendant
descList <- c(as.list(1:length(tree$tip.label)), descList)
# Convert `descList` to `edge_array` that matches the order of things in `tree$edge`
# For each entry in the tree$edge table, sum the descendants for each sample
# `tree$edge[i, 2]` is the node ID.
desc.idx <- mclapply(1:nrow(tree$edge), function(i) {descList[[tree$edge[i, 2]]]}, mc.cores=use.cores);
edge.array <- foreach(
	i = split(1:length(desc.idx), ceiling(seq_along(1:length(desc.idx))/100)),
	.combine='rbind',
	.multicombine=T) %dopar% {
		do.call(rbind, lapply(desc.idx[i], function(ii) {colSums(sp.o.t[ii, ,drop=F], na.rm=T)}))
	}
#Pre-calculate pre-req edge weights
edge.weight <- edge.array %*% diag(1/size.sample); colnames(edge.weight) <- colnames(edge.array);
#Get edge occurrence matrix (for unweighted UniFrac)
edge.occ <- (edge.array > 0) - 0;
edge.occ.list <- apply((edge.array > 0), 2, which);
#Get pre-multiplied edge.lengths per sample
edge.occ.len <- tree$edge.length * edge.occ;
#Get tree in "postorder" order
z = reorder.phylo(tree, order="postorder");
#Get "tip ages", or horizontal position of edges in re-ordered phylo object
tip.ages = ape_node_depth_edge_length(Ntip = length(tree$tip.label), Nnode = tree$Nnode, edge = z$edge, Nedge = nrow(tree$edge)[1], edge.length = z$edge.length);
#Keep only tips
tip.ages <- tip.ages[1:length(tree$tip.label)]
#Rename (when in doubt)
names(tip.ages) <- z$tip.label;
#Reorder
tip.ages <- tip.ages[rownames(o.t)];
########################

########################
#Calculate UniFrac distances based on pre-formed arrays
########################
#Unweighted UniFrac
Sys.time();
size.chunk <- 1000;
sim_list.uw <- foreach(i=split(1:length(spn), ceiling(seq_along(1:length(spn))/size.chunk))) %dopar% {
	unlist(lapply(spn[i], function(sp) {
		#Get occurrence vectors for current samples
		a <- edge.occ.list[[sp[1]]]; b <- edge.occ.list[[sp[2]]];
		curr.shared_edges <- intersect(a, b);
		curr.union_edges <- union(a, b);
		#Get current edge lengths
		curr.edge.len <- edge.occ.len[, sp, drop=F];
		#Return: 1 - (shared branch length) / (total branch length)
		#=> formulation as a distance
		1 - sum(curr.edge.len[curr.shared_edges, ], na.rm=TRUE) / sum(curr.edge.len[curr.union_edges, ], na.rm=TRUE)
	}));
}
Sys.time();
########################
#Weighted UniFrac
Sys.time();
sim_list.w <- foreach(i=split(1:length(spn), ceiling(seq_along(1:length(spn))/size.chunk))) %dopar% {
	unlist(lapply(spn[i], function(sp) {
		#Get current samples and total sample sizes
		a <- sp[1]; b <- sp[2];
		a.T <- size.sample[a]; b.T <- size.sample[b];
		#Calculate branchweights
		curr.bw <- abs(edge.weight[, a] - edge.weight[, b]);
		#Return: (unshared branch length) / (total branch length)
		#=> formulation as distance
		sum(tree$edge.length * curr.bw, na.rm = TRUE) / sum(tip.ages * (ot.rel[,a] + ot.rel[,b]), na.rm = TRUE)
	}));
}
Sys.time();
```

The main "tweaks" are:

- building the ```edge.array``` in parallel (using mclapply and/or foreach)
- building the ```edge.array``` seems to be faster if the (sparse) OTU table is a sparse ```Matrix``` object
- fast ```edge.weight``` pre-calculations, based on linear algebra
- edge occurrence matrix as pre-calculated list of OTU indices per samples (```edge.occ.list```); this skips many futile double-zero comparisons for unweighted UniFrac later in the loop
- set-based operations (```intersect``` and ```union```) for unweighted UniFrac
- out-of-loop pre-calculations (```edge.weight```) for weighted UniFrac
- most importantly, the ```foreach``` parallelization ran much faster (at least on my multicore system) when chunked (using ```lapply``` on a list of sample pairs inside the ```foreach``` loop)

My code is certainly still very clunky, but I thought I'd share it here so that you can play around with the ideas a bit. I didn't deeply benchmark performance but on my system (40 CPU multicore), the unweighted UniFrac ran in ~8min and weighted UniFrac in little over 1h for the dataset described above (~4,000 by ~30,000); in the original implementation, I didn't wait all the way for it to finish and killed it after several hours instead.

Anyway. I did not do a pull request as my code is certainly not ready for release within ```phyloseq``` (I even messed around with your variable names, as you notice). Still, it may give you some ideas for future releases :-)

Best,

Sebastian

**Comment by colinbrislawn** [comment]
Very nice! :+1:

**Comment by joey711** [comment]
Sebastian @defleury ,

Very cool!  I have a couple of immediate comments/questions about implementing this into phyloseq directly.

(1) Is the "bigmemory" library necessary in your implementation, or can this be a "Suggests" (optional) dependency and still make gains?
(2) Relatedly, do you know off-hand if the performance increase is because of better memory handling (avoided swapping -- making a copy of a large dataset for each parallel thread blows up the peak memory), or is there a pure computational speed-up as well? Most of my benchmarking up until now identified computational bottlenecks during specific tree traversal steps, and for the most part those are already accomplished in very efficient low-level C steps that are called internally (as you probably noticed). 
(3) The edge weight pre-computation sgtm, but I'll need to remind myself what the current approach does instead before commenting further. 
(4) What is the code to define `ot`? (the previously built sparse Matrix object reassigned to `sp.o.t`)? I'm a fan of the "Matrix" package and its well implemented sparse operations make a lot of sense for this problem, at least for the common case where the data is sparse. I think the default instantiation method might be fine, and if there's any sparse / non-sparse performance fork needed we could just do it via internal S4 dispatch.

Anyway, those are just some quick thoughts. Very cool. If it is possible to implement and test relatively soon, this could make it into the next release on BioC, just a few weeks away.

Thanks and looking forward to your feedback

**Comment by defleury** [comment]
Joey,

thanks for the prompt feedback! And sorry for not replying sooner: I was on a long weekend's holiday...

Here's to your questions:

(1) In fact, ```bigmemory``` is not needed at all for the code I posted. I had included it for an earlier (and less efficient) version of the ```foreach``` loops and then forgot to take it out again. Sorry about that!

(2) I am not sure. I believe that most speed-ups are indeed related to faster processing, not memory. For the early pre-processing steps, parallelisation reduced processing time significantly. Same applies for taking some redundant computations out of the ```foreach``` loops. But most importantly, ```foreach``` seems to be faster if you give it chunks of UniFrac computations (```unlist(lapply())``` constructs within the loop), rather than individual tasks that generate significant overhead. But the use of sparse ```Matrix``` OTU tables seems to help, too, for ```colSums``` etc.

(3) According to your [distance-methods.R code](https://github.com/joey711/phyloseq/blob/0e719e23e8bd6b2b25f9a8c2a71d1621bf854f17/R/distance-methods.R), the original code for the edge weights is a vanilla ```for``` loop:
```R
# Convert `descList` to `edge_array` that matches the order of things in `tree$edge`
	edge_array <- matrix(0, nrow=nrow(tree$edge), ncol=nsamples(physeq), 
	                      dimnames=list(NULL, sample_names=sample_names(physeq)))
	for(i in 1:nrow(tree$edge)){
# For each entry in the tree$edge table, sum the descendants for each sample
# `tree$edge[i, 2]` is the node ID.
	  edge_array[i, ] <- colSums(OTU[descList[[tree$edge[i, 2]]], , drop=FALSE], na.rm = TRUE)
	}
```

(4) I called ```Matrix(otu.table, sparse=T)``` on the original (R vanilla ```matrix```) OTU table to make it explicitly sparse. However, according to the ```Matrix``` package [documentation](https://stat.ethz.ch/R-manual/R-devel/library/Matrix/html/Matrix.html), it should be sparse by default (as generally, more than half of the entries in the OTU table will be 0).

I didn't systematically benchmark the original ```fastUniFrac``` code to identify memory/CPU bottlenecks; I really just tried to get it to run a bit faster for my concrete dataset. I am quite sure that my suggested code is far from optimal, and many more tweaks/more elegant solutions are possible. But as phyloseq is such a nice tool and is used by more and more people, I thought I'd share my code here :-)

Best,


Sebastian

**Comment by defleury** [comment]
Joey,

I just found two bugs in the code I posted above (though your original code took care of it). The first one is pretty easy to fix: I forgot to include a line to match OTU table ```rownames``` to the edge names in the ```tree```. Including the following line (modified from your original function) in the code early on fixes that:

```R
if( !all(rownames(sp.o.t) == taxa_names(tree)) ){sp.o.t <- sp.o.t[taxa_names(tree), ]; o.t <- o.t[taxa_names(tree), ]}
```



The second bug is a bit trickier: I made a mistake in the unweighted UniFrac code. When calculating the numerator and denominator, I was summing over some branches twice in this line of code:

```R
1 - sum(curr.edge.len[curr.shared_edges, ], na.rm=TRUE) / sum(curr.edge.len[curr.union_edges, ], na.rm=TRUE)
```

The correct line would be:
```R
1 - sum(tree$edge.length[curr.shared_edges], na.rm=TRUE) / sum(tree$edge.length[curr.union_edges], na.rm=TRUE)
```

However, the error introduced in this way was so small for my original test set that I didn't detect it when comparing to your results. It only popped out now, with a new dataset I was trying.

Now, however, the code really gives the same results as yours :-)

**Comment by joey711** [comment]
Sebastian,

Thanks for following up. This is great. I won't have time to merge into the next BioC release (in the next week or so), but I will merge into master as soon as I can, and it will jump into the following release.

Btw, if you think your test data is a useful addition to the unit tests already included, would be helpful to include that as an additional unit test as well. Just a thought.

**Comment by benjjneb** [comment]
While we're speeding up phyloseq's UniFrac, the following is a drop in replacement for the pre-computation of the edge-weights that speeds up computation about 10x in the Global patterns dataset (20s -> 2s on my laptop) for identical results:

```R
  ########################################
  # Build the requisite matrices as defined 
  # in the Fast UniFrac article.
  ########################################
  ## This only needs to happen once in a call to UniFrac.
  ## Notice that A and B do not appear in this section.
  # Begin by building the edge descendants matrix (edge-by-sample)
  # `edge_array`
  #
### BEGINNING OF CHANGES
  # Create a list of descendants, starting from the first internal node (root)
  ntip <- length(tree$tip.label)
  if(ntip != ntaxa(physeq)) stop("Incompatible tree and OTU table!")
  # Create a matrix that maps each internal node to its 2 descendants
  # This matrix doesn't include the tips, so must use node#-ntip to index into it
  node.desc <- matrix(tree$edge[order(tree$edge[,1]),][,2],byrow=TRUE,ncol=2)
  # Define the edge_array object
  # Right now this is a node_array object, each row is a node (including tips)
  # It will be subset and ordered to match tree$edge later
  edge_array <- matrix(0, nrow=ntip+tree$Nnode, ncol=nsamples(physeq), 
                       dimnames=list(NULL, sample_names=sample_names(physeq)))
  # Load the tip counts in directly
  edge_array[1:ntip,] <- OTU
  # Get a list of internal nodes ordered by increasing depth
  ord.node <- order(node.depth(tree))[(ntip+1):(ntip+tree$Nnode)]
  # Loop over internal nodes, summing their descendants to get that nodes count
  for(i in ord.node){
    edge_array[i,] <- colSums(edge_array[node.desc[i-ntip,], , drop=FALSE], na.rm = TRUE)
  }
  # Keep only those with a parental edge (drops root) and order to match tree$edge
  edge_array <- edge_array[tree$edge[,2],]
  # Remove unneeded variables.
  rm(node.desc)
### END OF CHANGES
```

The key here is that the fastUnifrac algorithm wasn't really implemented previously,  `descList <- prop.part(tree, check.labels = FALSE)` returns a list of all tips below each node, thus for large trees there can be many, many tips that are then all summed (in the colSums call) to get the counts for the associated edge. That approach scales like ~ntaxa^2. The above saves just the two descendant nodes for each node, and only sums those two nodes to get the associated sample-wise counts. That scales ~ntaxa

This cant just be parallelized, because loop calculations depend on previous iterations of the loop. But the performance gain scales as ~ntaxa.

**Comment by joey711** [comment]
Thanks, Ben, I'll make sure this gets in as well. Or you can make it a proper PR. Depends how fast you want it in there O:-)

**Comment by joey711** [comment]
Closed.
